File: SqlOrganize/Sql/Logging.cs
using Microsoft.VisualBasic;
using Newtonsoft.Json;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;

namespace SqlOrganize.Sql
{
    /// <summary>
    /// Registro de errores
    /// </summary>
    public class Logging
    {
        public enum Level
        {
            Success,
            Info,
            Warning,
            Error,
        }

        /// <summary> Dictionary to store logs grouped by key. </summary>
        /// <example>
        /// [
        ///   "asignatura": [  {"level": LEVEL_ERROR, "msg": "No puede estar vacío", type: "required"}  ]
        ///   "plan": [ {"level": LEVEL_WARNING, "msg: "No tiene cargas horarias asociadas", type: "user"}  ]
        ///   "numero": [
        ///       { "level": LEVEL_ERROR, "msg": "No es unico", type: "not_unique"}
        ///       { "level": LEVEL_WARNING, "msg": "Esta fuera del rango permitido", type: "out_of_range"}
        ///   ]
        /// ]
        /// </example>
        public Dictionary<string, List<(Level Level, string Message, string? Type)>> Logs { get; } = new ();


        /// <summary> Get logs for a specific key. </summary>
        public List<(Level Level, string Message, string? Type)>? GetLogsByKey(string key) =>
            Logs.TryGetValue(key, out var logs) ? logs : null;

        /// <summary>  Clears logs for a specific key. </summary>
        public void ClearByKey(string key) => Logs.Remove(key);

        /// <summary> Clears all logs. </summary>
        public void Clear() => Logs.Clear();

        /// <summary> Adds a new log entry. </summary>
        public void AddLog(string key, string message, string? type = null, Level level = Level.Info)
        {
            if (!Logs.ContainsKey(key))
                Logs[key] = new();

            Logs[key].Add((level, message, type));
            // Sorting logs by level
            Logs[key] = Logs[key].OrderBy(log => log.Level).ToList();
        }

        /// <summary>  Merges logs from another Logging instance. </summary>
        public void Merge(Logging other)
        {
            foreach (var (key, otherLogs) in other.Logs)
            {
                if (!Logs.ContainsKey(key))
                    Logs[key] = new();

                Logs[key].AddRange(otherLogs);
                Logs[key] = Logs[key].OrderBy(log => log.Level).ToList();
            }
        }

        /// <summary> Adds an error log entry. </summary>
        public void AddErrorLog(string key, string message, string? type = null) =>
            AddLog(key, message, type, Level.Error);

        /// <summary> Retrieves the highest log level for a specific key. </summary>
        public Level? GetHighestLevelForKey(string key) =>
            Logs.TryGetValue(key, out var logs) ? logs.Max(log => log.Level) : null;

        /// <summary> Checks if any logs exist. </summary>
        public bool HasLogs() => Logs.Any();

        /// <summary>  Checks if any error logs exist. </summary>
        public bool HasErrors() =>
            Logs.Values.Any(logList => logList.Any(log => log.Level == Level.Error));

        /// <summary> Retrieves logs filtered by a specific level. </summary>
        public Dictionary<string, List<(Level Level, string Message, string? Type)>> GetLogsByLevel(Level level) =>
            Logs
                .Where(entry => entry.Value.Any(log => log.Level == level))
                .ToDictionary(
                    entry => entry.Key,
                    entry => entry.Value.Where(log => log.Level == level).ToList()
                );


        /// <summary> Converts logs to a JSON string representation. </summary>
        public override string ToString() =>
            JsonConvert.SerializeObject(
                Logs.SelectMany(entry => entry.Value.Select(log => new
                {
                    Key = entry.Key,
                    log.Message
                })),
                Formatting.Indented
            );

    }
}

File: SqlOrganize/Sql/Logging.php
<?php

namespace SqlOrganize\Sql;

/**
 * Registro de errores
 */
class Logging
{
    const LEVEL_SUCCESS = 0;
    const LEVEL_INFO = 1;
    const LEVEL_WARNING = 2;
    const LEVEL_ERROR = 3;

    /**
     * Dictionary to store logs grouped by key.
     * @example
     * [
     *   "asignatura" => [
     *       ["level" => LEVEL_ERROR, "message" => "No puede estar vacío", "type" => "required"]
     *   ],
     *   "plan" => [
     *       ["level" => LEVEL_WARNING, "message" => "No tiene cargas horarias asociadas", "type" => "user"]
     *   ],
     *   "numero" => [
     *       ["level" => LEVEL_ERROR, "message" => "No es unico", "type" => "not_unique"],
     *       ["level" => LEVEL_WARNING, "message" => "Esta fuera del rango permitido", "type" => "out_of_range"]
     *   ]
     * ]
     */
    private array $logs = [];

    /**
     * Get logs for a specific key.
     */
    public function getLogsByKey(string $key): ?array
    {
        return $this->logs[$key] ?? null;
    }

    /**
     * Clears logs for a specific key.
     */
    public function clearByKey(string $key): void
    {
        unset($this->logs[$key]);
    }

    /**
     * Clears all logs.
     */
    public function clear(): void
    {
        $this->logs = [];
    }

    /**
     * Adds a new log entry.
     */
    public function addLog(string $key, string $message, ?string $type = null, int $level = self::LEVEL_INFO): void
    {
        if (!isset($this->logs[$key])) {
            $this->logs[$key] = [];
        }

        $this->logs[$key][] = [
            'level' => $level,
            'message' => $message,
            'type' => $type
        ];

        // Sorting logs by level
        usort($this->logs[$key], function($a, $b) {
            return $a['level'] <=> $b['level'];
        });
    }

    /**
     * Merges logs from another Logging instance.
     */
    public function merge(Logging $other): void
    {
        foreach ($other->logs as $key => $otherLogs) {
            if (!isset($this->logs[$key])) {
                $this->logs[$key] = [];
            }

            $this->logs[$key] = array_merge($this->logs[$key], $otherLogs);
            
            // Sorting logs by level
            usort($this->logs[$key], function($a, $b) {
                return $a['level'] <=> $b['level'];
            });
        }
    }

    /**
     * Adds an error log entry.
     */
    public function addErrorLog(string $key, string $message, ?string $type = null): void
    {
        $this->addLog($key, $message, $type, self::LEVEL_ERROR);
    }

    /**
     * Retrieves the highest log level for a specific key.
     */
    public function getHighestLevelForKey(string $key): ?int
    {
        if (!isset($this->logs[$key]) || empty($this->logs[$key])) {
            return null;
        }

        return max(array_column($this->logs[$key], 'level'));
    }

    /**
     * Checks if any logs exist.
     */
    public function hasLogs(): bool
    {
        return !empty($this->logs);
    }

    /**
     * Checks if any error logs exist.
     */
    public function hasErrors(): bool
    {
        foreach ($this->logs as $logList) {
            foreach ($logList as $log) {
                if ($log['level'] === self::LEVEL_ERROR) {
                    return true;
                }
            }
        }
        return false;
    }

    /**
     * Retrieves logs filtered by a specific level.
     */
    public function getLogsByLevel(int $level): array
    {
        $result = [];
        
        foreach ($this->logs as $key => $logList) {
            $filteredLogs = array_filter($logList, function($log) use ($level) {
                return $log['level'] === $level;
            });
            
            if (!empty($filteredLogs)) {
                $result[$key] = array_values($filteredLogs);
            }
        }
        
        return $result;
    }

    /**
     * Converts logs to a JSON string representation.
     */
    public function __toString(): string
    {
        $flattened = [];
        
        foreach ($this->logs as $key => $logList) {
            foreach ($logList as $log) {
                $flattened[] = [
                    'key' => $key,
                    'message' => $log['message']
                ];
            }
        }
        
        return json_encode($flattened, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE);
    }

    /**
     * Get all logs (for debugging or direct access)
     */
    public function getAllLogs(): array
    {
        return $this->logs;
    }
}

?>